### PR TITLE
{AKS} Add provisioningState retry logic with two-phase check validation

### PR DESCRIPTION
**Pull request by deveshdama**
Add opt-in retry logic to the Azure CLI test SDK for handling provisioningState race conditions caused by external modifications (e.g., Azure Policy) during live tests.

**Related command**
N/A

**Description**

### Problem:
provisioningState race condition caused by Azure Policy seen in AKS CLI Runners that use azure cli test sdk. Live tests in Microsoft-managed subscriptions intermittently fail with:

`Query 'provisioningState' doesn't yield expected value 'Succeeded', instead the actual value is 'Updating'`

| Time | Actor | Action | provisioningState |
|------|-------|--------|-------------------|
| 20:46:01 | Azure CLI | PUT managed cluster (cli test) | - |
| 20:56:52 | AKS | PUT completes successfully | Succeeded |
| 20:57:08 | Azure Policy | PUT managed cluster (azure policy compliance) | Updating |
| 20:57:14 | Azure CLI SDK | GET managed cluster (test assertion) | Updating (cli test failure) |

Root cause (thanks @FumingZhang for the thorough investigation):

The failure occurs due to a race condition between the Azure CLI SDK's internal GET request and an Azure Policy PUT request(in our case).
Sequence of events causing this race:

CLI sends PUT to create/update a managed cluster - the operation succeeds (`provisioningState = Succeeded` at T+0).
Azure Policy sends its own PUT to the same resource at T+6s (`user agent = ARM`), re-applying policy compliance. This resets provisioningState to Updating.
CLI SDK sends GET at T+8s to read the resource and run test assertions. It sees `provisioningState = Updating` instead of `Succeeded`, and the test fails - even though the CLI operation itself succeeded.

### Proposed Fix in the PR:
When env var `AZURE_CLI_TEST_RETRY_PROVISIONING_CHECK=true`, test framework `cmd()` method splits the checks into two sequential phases:

Phase-1 : only provisioningState validation : If `provisioningState != Succeeded`, poll the resource via `az resource show --ids <id>` with exponential backoff until it reaches a terminal state `Succeeded`, `Failed`, or `Canceled`.
Phase-2 : remaining validations : Validate all non-provisioningState checks against the original command result, validates CLI operation itself produced correct output.

### usage of etag:

When provisioningState is not `Succeeded`, poll the resource until it reaches a terminal state or timeout. Expectation here is that extrenal operation completes and provisioningState returns `Succeeded`, allowing the tests to succeed.

`etag` here is being used for observability only and it does not change the flow of the code.
1. Original etag is captured during initial command's execution result.
2. If provisioningState does not match with the expected value, polling begins:
      - On each poll, compare current etag on the resource from last seen value.
      - if they differ, etag has been changed by an external actor(in our case azure policy). Log a warning.
3. On timeout, both original and current etags are logged in error message for debugging

**Testing Guide**
Attaching testing script. https://github.com/Azure/azure-cli/pull/32939#issuecomment-4026221613

**History Notes**

[Component Name 1] BREAKING CHANGE: `az command a`: Make some customer-facing breaking change
[Component Name 2] `az command b`: Add some customer-facing feature

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Thank you for your contribution! We will review the pull request and get back to you soon.

**Comment by deveshdama** [comment]
[repro_provisioning_retry.py](https://github.com/user-attachments/files/25851758/repro_provisioning_retry.py)

Attaching repro script for validations.

**Comment by deveshdama** [comment]
> The changes look good to me based on testing in live mode. However, will the recording file generated during live mode execution, especially when the provisioning state check is retried, still work correctly in replay mode? I’m concerned that there could be additional resource get operations recorded, which might cause a mismatch in replay mode.

The retry logic is gated behind self.is_live, so it never runs during playback.
Also since the feature is opt-in and only activates when env var `AZURE_CLI_TEST_RETRY_PROVISIONING_CHECK=true` so test recording/playback workflows are unaffected.

cc: @FumingZhang

**Comment by FumingZhang** [comment]
> The retry logic is gated behind self.is_live, so it never runs during playback.

This is exactly what concerns me. In replay mode, it shouldn't need to send a resource get request using `az resource show` after the resource is created. However, that request would be recorded in the cassette file if the test running in live mode encountered an issue and retried. In such cases, when the vcr library tries to match the request and response from the cassette file during replay testing, there could be a mismatch. Since some request/response pairs can be ignored, I would like your help to confirm whether this change could cause the newly generated cassette file to fail a subsequent replay test,  @deveshdama

**Comment by deveshdama** [comment]
> > The retry logic is gated behind self.is_live, so it never runs during playback.
> 
> This is exactly what concerns me. In replay mode, it shouldn't need to send a resource get request using `az resource show` after the resource is created. However, that request would be recorded in the cassette file if the test running in live mode encountered an issue and retried. In such cases, when the vcr library tries to match the request and response from the cassette file during replay testing, there could be a mismatch. Since some request/response pairs can be ignored, I would like your help to confirm whether this change could cause the newly generated cassette file to fail a subsequent replay test, @deveshdama

The retry is gated behind `if self.is_live`. During live recording, extra poll requests get recorded
into the cassette. During replay, the retry is skipped - those entries sit unconsumed. VCR is strict
on missing entries (fails if a request has no match) but lenient on extra entries (silently ignores
unconsumed ones). The only side effect is a slightly larger yaml file.

I verified this by tracing through vcrpy's source:
- [__exit__](https://github.com/kevin1024/vcrpy/blob/v8.1.1/vcr/cassette.py#L86-L99) only calls `_save()` + unpatcher cleanup - no `all_played` check
- [_save()](https://github.com/kevin1024/vcrpy/blob/v8.1.1/vcr/cassette.py#L349-L357) with default [drop_unused_requests=False](https://github.com/kevin1024/vcrpy/blob/v8.1.1/vcr/cassette.py#L180) calls [_as_dict()](https://github.com/kevin1024/vcrpy/blob/v8.1.1/vcr/cassette.py#L338-L339) which returns all entries
unconditionally
- [all_played](https://github.com/kevin1024/vcrpy/blob/v8.1.1/vcr/cassette.py#L209-L212) is defined but never called by vcrpy itself

this script demonstrates this:
https://gist.github.com/deveshdama/4a94fb462660731170d9548cd1df5c1b

```bash
pip install vcrpy && python3 vcr_proof_demo.py
```
Output:
```console
PHASE 2: REPLAY  (record_mode='once', only 1 request, 2 skipped)
   Consumed: 1/3 interactions
   Unconsumed: 2  (the retry poll entries)
   cassette.all_played = False
   >>> VCR exited cleanly. No error from unconsumed entries. <<<
```

cc: @FumingZhang

File: src/azure-cli-testsdk/azure/cli/testsdk/base.py [hunk 1]
```diff
@@ -78,6 +80,129 @@ def is_empty(self):  # pylint: disable=no-self-use
...
+                return True
+        return False
+
+    def _cmd_with_retry(self, command, checks, cli_ctx):
+        """Execute command with two-phase check validation.
+
+        Phase 1: Validate provisioningState == Succeeded, retrying with poll if needed.
+        Phase 2: Validate all remaining checks on the original result.
+
+        Uses etag to detect external modifications (e.g. Azure Policy).
```
Comment (jiasli) [inline src/azure-cli-testsdk/azure/cli/testsdk/base.py:115]: There is a lot of code depending on etag. It's better to explain what the logic is in the PR description.

**Comment by jiasli** [comment]
1. It's better to explain the **`provisioningState` race conditions** in detail (in a step by step manner) to help others understand what the problem actually is.

2. If I understand correctly, nowadays, running live tests in Microsoft-managed subscriptions is not a standard approach. This PR may mitigate some problems for a while, but when new policies are deployed, other issues may emerge. The standard is to use BAMI tenants and subscriptions.

3. Minor: It's better to use the correct [Markdown syntax](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax) in the PR description to make it more readable.

**Comment by jiasli** [comment]
Claude Opus 4.6 states this PR shows strong AI aiding signals. If it is true, it may be worth considering if we really need this level of complexity (over engineering?), as this will make reviewing and maintaining harder and require more resources for testing.

Pasting Claude Opus 4.6's answer below:

---

Based on analyzing the PR's code, commit messages, PR description, and overall patterns, there are **several strong signals suggesting this PR was significantly aided by AI**:

---

### 🔍 Evidence

#### 1. **Highly polished commit message structure**
The single commit message reads like a well-formatted AI-generated summary — bullet points with marketing-style feature descriptions:
> - *"Opt-in via environment variable (zero risk to existing tests)"*
> - *"Exponential backoff with jitter (configurable via env vars)"*
> - *"Terminal state short-circuit (Failed/Canceled stop polling)"*

This level of structured, polished prose in a commit message is uncommon for typical human-written commits, which tend to be terser.

#### 2. **PR description ≈ commit message ≈ code docstrings**
The PR description, commit message body, and inline code docstrings all repeat nearly identical phrasing word-for-word (e.g., "two-phase check approach", "Phase 1/Phase 2" descriptions). AI tools tend to generate consistent descriptions across artifacts from the same prompt.

#### 3. **Test file is suspiciously comprehensive for a single commit**
The test file (`test_provisioning_retry.py`) is **406 lines** with 13 well-structured test cases covering every edge case (happy path, timeout, terminal failure, no resource ID, ETag change, two-phase combinations). This level of test coverage in a single commit — with consistent naming conventions, detailed docstrings for every test, and clean mock patterns — is characteristic of AI-generated test code.

#### 4. **Boilerplate docstring style**
Every method has a precisely formatted docstring:
```python
def _cmd_with_retry(self, command, checks, cli_ctx):
    """Execute command with two-phase check validation.

    Phase 1: Validate provisioningState == Succeeded, retrying with poll if needed.
    Phase 2: Validate all remaining checks on the original result.

    Uses etag to detect external modifications (e.g. Azure Policy).
    """
```
This is clean but feels templated — the style is more consistent than typical human code where docstring quality tends to vary.

#### 5. **PR description has unfilled boilerplate**
The history notes section was left as template text:
> `[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change`

This is common when AI generates the PR body from a template prompt but the author doesn't review/clean up the parts that need manual input.

#### 6. **Copilot reviewer suggestion was adopted verbatim**
The `_is_provisioning_state_check` method in the latest commit matches the Copilot reviewer's [suggested code](https://github.com/Azure/azure-cli/pull/32939#discussion_r2904049637) almost exactly, which is fine (that's the purpose of suggestions), but combined with other signals, it reinforces the AI-assisted pattern.

#### 7. **Code and test style uniformity**
The production code and test code show unusually consistent style — same commenting patterns, same level of detail, same structural approach. Human-written code typically shows subtle style shifts between production and test code.

---

### 📊 Assessment

| Signal | Strength |
|--------|----------|
| Commit message polish | 🟡 Medium-High |
| Description/docstring repetition | 🟡 Medium |
| Test coverage completeness in single commit | 🔴 High |
| Boilerplate left unfilled | 🟡 Medium |
| Overall code uniformity | 🟡 Medium |

**Verdict: Likely AI-assisted** — The code quality is decent and the logic is sound, but the uniformity of style, the comprehensiveness of tests in a single pass, and the polished-but-template-like documentation strongly suggest significant AI involvement in generating the code and tests. The author likely used an AI tool (possibly GitHub Copilot or ChatGPT) to scaffold the implementation and tests, then made targeted adjustments.

> **Note:** AI-assisted code isn't inherently bad — the code itself is functional and well-tested. The concern is more about whether the author fully understands every nuance (e.g., the replay-mode question raised by @FumingZhang, or your question about the ETag logic explanation).

File: src/azure-cli-testsdk/azure/cli/testsdk/base.py (hunk 1 above)
Comment (deveshdama) [reply src/azure-cli-testsdk/azure/cli/testsdk/base.py:115]: Added a clear PR description.

**Comment by deveshdama** [comment]
> 1. It's better to explain the **`provisioningState` race conditions** in detail (in a step by step manner) to help others understand what the problem actually is.
> 2. If I understand correctly, nowadays, running live tests in Microsoft-managed subscriptions is not a standard approach. This PR may mitigate some problems for a while, but when new policies are deployed, other issues may emerge. The standard is to use BAMI tenants and subscriptions.
> 3. Minor: It's better to use the correct [Markdown syntax](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax) in the PR description to make it more readable.

1. I've tried to explain the issue and the fix in detail in the PR description. Added a table explaining the exact sequence from one of the failures we investigated.
2. @FumingZhang can help claify this one.
3. Done, thanks for pointing it out.

**Comment by deveshdama** [comment]
> Claude Opus 4.6 states this PR shows strong AI aiding signals. If it is true, it may be worth considering if we really need this level of complexity (over engineering?), as this will make reviewing and maintaining harder and require more resources for testing.
> 
> Pasting Claude Opus 4.6's answer below:
> 
> Based on analyzing the PR's code, commit messages, PR description, and overall patterns, there are **several strong signals suggesting this PR was significantly aided by AI**:
> 
> ### 🔍 Evidence
> #### 1. **Highly polished commit message structure**
> The single commit message reads like a well-formatted AI-generated summary — bullet points with marketing-style feature descriptions:
> 
> > * _"Opt-in via environment variable (zero risk to existing tests)"_
> > * _"Exponential backoff with jitter (configurable via env vars)"_
> > * _"Terminal state short-circuit (Failed/Canceled stop polling)"_
> 
> This level of structured, polished prose in a commit message is uncommon for typical human-written commits, which tend to be terser.
> 
> #### 2. **PR description ≈ commit message ≈ code docstrings**
> The PR description, commit message body, and inline code docstrings all repeat nearly identical phrasing word-for-word (e.g., "two-phase check approach", "Phase 1/Phase 2" descriptions). AI tools tend to generate consistent descriptions across artifacts from the same prompt.
> 
> #### 3. **Test file is suspiciously comprehensive for a single commit**
> The test file (`test_provisioning_retry.py`) is **406 lines** with 13 well-structured test cases covering every edge case (happy path, timeout, terminal failure, no resource ID, ETag change, two-phase combinations). This level of test coverage in a single commit — with consistent naming conventions, detailed docstrings for every test, and clean mock patterns — is characteristic of AI-generated test code.
> 
> #### 4. **Boilerplate docstring style**
> Every method has a precisely formatted docstring:
> 
> ```python
> def _cmd_with_retry(self, command, checks, cli_ctx):
>     """Execute command with two-phase check validation.
> 
>     Phase 1: Validate provisioningState == Succeeded, retrying with poll if needed.
>     Phase 2: Validate all remaining checks on the original result.
> 
>     Uses etag to detect external modifications (e.g. Azure Policy).
>     """
> ```
> 
> This is clean but feels templated — the style is more consistent than typical human code where docstring quality tends to vary.
> 
> #### 5. **PR description has unfilled boilerplate**
> The history notes section was left as template text:
> 
> > `[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change`
> 
> This is common when AI generates the PR body from a template prompt but the author doesn't review/clean up the parts that need manual input.
> 
> #### 6. **Copilot reviewer suggestion was adopted verbatim**
> The `_is_provisioning_state_check` method in the latest commit matches the Copilot reviewer's [suggested code](https://github.com/Azure/azure-cli/pull/32939#discussion_r2904049637) almost exactly, which is fine (that's the purpose of suggestions), but combined with other signals, it reinforces the AI-assisted pattern.
> 
> #### 7. **Code and test style uniformity**
> The production code and test code show unusually consistent style — same commenting patterns, same level of detail, same structural approach. Human-written code typically shows subtle style shifts between production and test code.
> 
> ### 📊 Assessment
> Signal	Strength
> Commit message polish	🟡 Medium-High
> Description/docstring repetition	🟡 Medium
> Test coverage completeness in single commit	🔴 High
> Boilerplate left unfilled	🟡 Medium
> Overall code uniformity	🟡 Medium
> **Verdict: Likely AI-assisted** — The code quality is decent and the logic is sound, but the uniformity of style, the comprehensiveness of tests in a single pass, and the polished-but-template-like documentation strongly suggest significant AI involvement in generating the code and tests. The author likely used an AI tool (possibly GitHub Copilot or ChatGPT) to scaffold the implementation and tests, then made targeted adjustments.
> 
> > **Note:** AI-assisted code isn't inherently bad — the code itself is functional and well-tested. The concern is more about whether the author fully understands every nuance (e.g., the replay-mode question raised by @FumingZhang, or your question about the ETag logic explanation).


@jiasli 
copilot has been an assistant with most of the work. We've been looking at this issue for some time now and it took some thorough investigation from @FumingZhang to get to the rootcause. I can share some kusto/queries results if that helps. This is an intermittent issue that generates a lot of false positives in our test runs.
Last 30 days stats:
FailureCount : 204
DistinctTests : 85

over-engineering concern - `_cmd_with_retry` is about 60 lines, the rest is just tests and guard checks. Since this touches the test framework itself we wanted good coverage, don't want to introduce a bug here. Happy to make changes if you have additional feedback on improving this. This is an opt-in feature and is turned off by default.

On whether we need this - @FumingZhang can comment on the BAMI tenant migration as the long-term fix here. For now, my team has been ignoring all failures with this error message, could cause us to miss out on actual failures.

**Comment by FumingZhang** [comment]
BAMI tenant seems like a viable option, but after looking into it further, I discovered it has a strict 90-day lifespan. This makes it more appropriate for temporary tests rather than ongoing tests meant to monitor code quality. Additionally, migrating from one tenant to another after expiration could be cumbersome.
